hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotDeleteRequest.java

```diff
@@ -0,0 +1,217 @@
+    // Verify snapshot name. TODO: Can remove
+    OmUtils.validateSnapshotName(snapshotName);
+
+    String volumeName = deleteSnapshotRequest.getVolumeName();
+    String bucketName = deleteSnapshotRequest.getBucketName();
+
+    // Permission check
+    UserGroupInformation ugi = createUGI();
+    String bucketOwner = ozoneManager.getBucketOwner(volumeName, bucketName,
+        IAccessAuthorizer.ACLType.READ, OzoneObj.ResourceType.BUCKET);
+    if (!ozoneManager.isAdmin(ugi) &&
+        !ozoneManager.isOwner(ugi, bucketOwner)) {
+      throw new OMException(
+          "Only bucket owners and Ozone admins can delete snapshots",
+          OMException.ResultCodes.PERMISSION_DENIED);
+    }
+
+    // Set deletion time here so OM leader and follower would have the
+    // exact same timestamp.
+    OMRequest.Builder omRequestBuilder = omRequest.toBuilder()
+        .setDeleteSnapshotRequest(
+            DeleteSnapshotRequest.newBuilder()
+                .setVolumeName(volumeName)
+                .setBucketName(bucketName)
+                .setSnapshotName(snapshotName)
+                .setDeletionTime(Time.now()));
+
+    return omRequestBuilder.build();
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long transactionLogIndex,
+      OzoneManagerDoubleBufferHelper ozoneManagerDoubleBufferHelper) {
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumSnapshotDeletes();
+
+//    boolean acquiredBucketLock = false;
+    boolean acquiredSnapshotLock = false;
+    IOException exception = null;
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+
+    OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(
+        getOmRequest());
+    OMClientResponse omClientResponse = null;
+    AuditLogger auditLogger = ozoneManager.getAuditLogger();
+
+    UserInfo userInfo = getOmRequest().getUserInfo();
+
+    final DeleteSnapshotRequest request =
+        getOmRequest().getDeleteSnapshotRequest();
+
+    final String volumeName = request.getVolumeName();
+    final String bucketName = request.getBucketName();
+    final String snapshotName = request.getSnapshotName();
+    final long deletionTime = request.getDeletionTime();
+
+    SnapshotInfo snapshotInfo = null;
+
+    try {
+      // TODO: Do we need a bucket lock here? Probably not?
+//      acquiredBucketLock =
+//          omMetadataManager.getLock().acquireReadLock(BUCKET_LOCK,
+//              volumeName, bucketName);
+
+      acquiredSnapshotLock =
+          omMetadataManager.getLock().acquireWriteLock(SNAPSHOT_LOCK,
+              volumeName, bucketName, snapshotName);
+
+      // Retrieve SnapshotInfo from the table
+      String tableKey = SnapshotInfo.getTableKey(volumeName, bucketName,
+          snapshotName);
+      snapshotInfo =
+          omMetadataManager.getSnapshotInfoTable().get(tableKey);
+
+      if (snapshotInfo == null) {
```

Comment:
The previous check is done in `preExecute`. Thus we have to check it again here in `validateAndUpdateCache` because stages of different requests can interleave.